## Fun With Books EDA
Doing some exploration of the data to see what might be interesting to display in my product. 

### Set-Up
Loading requisite packages and the data itself. 

In loading the dataset, there were a couple of issues with comma delimiters in author names. I made a couple of manual adjustments to the `books.csv` file, which are now stored as such. 

In [22]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import plotnine as pn

In [12]:
data = pd.read_csv('books.csv')
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11127 entries, 0 to 11126
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   bookID              11127 non-null  int64  
 1   title               11127 non-null  object 
 2   authors             11127 non-null  object 
 3   average_rating      11127 non-null  float64
 4   isbn                11127 non-null  object 
 5   isbn13              11127 non-null  int64  
 6   language_code       11127 non-null  object 
 7   num_pages           11127 non-null  int64  
 8   ratings_count       11127 non-null  int64  
 9   text_reviews_count  11127 non-null  int64  
 10  publication_date    11127 non-null  object 
 11  publisher           11127 non-null  object 
dtypes: float64(1), int64(5), object(6)
memory usage: 1.0+ MB


11127 entries with no apparent null values!

### Univariate Exploration

In [16]:
# function to facet a set of variables together for univariate or bivariate 
def melt_plot(x, y, data, title, scales = "free", geom_type = pn.geom_point()):
    if y is not None:
        id_vars = ["datetime",y]
        aes_set = pn.aes("value", y)
    else:
        id_vars = ["datetime"]
        aes_set = pn.aes("value")

    data_long = data.melt(id_vars=id_vars, value_vars=x)
    plot = (
        pn.ggplot(data_long, aes_set) 
        + geom_type
        + pn.facet_wrap("~variable", scales=scales)   
        + pn.labs(title=title)
        )   
    return plot

#### Numeric Data
There are just a couple of numeric variables found in this dataset, taking a quick look at them. 

In [18]:
numeric = ['average_rating','ratings_count','text_reviews_count','num_pages']
data[numeric].describe()

average_rating  ratings_count  text_reviews_count     num_pages
count    11127.000000   1.112700e+04        11127.000000  11127.000000
mean         3.933631   1.793641e+04          541.854498    336.376921
std          0.352445   1.124794e+05         2576.176608    241.127305
min          0.000000   0.000000e+00            0.000000      0.000000
25%          3.770000   1.040000e+02            9.000000    192.000000
50%          3.960000   7.450000e+02           46.000000    299.000000
75%          4.135000   4.993500e+03          237.500000    416.000000
max          5.000000   4.597666e+06        94265.000000   6576.000000

In [26]:
num_pivot = data.melt(id_vars="bookID", value_vars=numeric)
num_pivot.head()
(
    pn.ggplot(num_pivot, pn.aes(x="value")) +
    pn.geom_histogram() +
    pn.facet_wrap("~variable", scales="free")
)

/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/plotnine/stats/stat_bin.py:109: PlotnineWarning: 'stat_bin()' using 'bins = 195442'. Pick better value with 'binwidth'.
